### PR TITLE
Improvement on merge task performance

### PR DESCRIPTION
**Pull request by brnleehng**
Added tasks that will merge into chunks

Should we add an optional argument for the size of the merge worker?

### CODE REVIEW

File: inst/startup/merger.R [hunk 1]
```diff
@@ -22,6 +17,16 @@ batchTaskId <- Sys.getenv("AZ_BATCH_TASK_ID")
 batchJobPreparationDirectory <-
   Sys.getenv("AZ_BATCH_JOB_PREP_WORKING_DIR")
 batchTaskWorkingDirectory <- Sys.getenv("AZ_BATCH_TASK_WORKING_DIR")
+taskPackageDirectory <- paste0(batchTaskWorkingDirectory)
+
+libPaths <- c(
+  taskPackageDirectory,
+  jobPrepDirectory,
+  "/mnt/batch/tasks/shared/R/packages",
```
Comment (paselem) [inline inst/startup/merger.R]: this should probably be:
```R
clusterPackageDirectory <- paste0(Sys.getenv("AZ_BATCH_NODE_SHARED_DIR", "/R/packages")

libPaths <- c(
...
clusterPacakgeDirectory,
...
)
```

File: inst/startup/merger.R [hunk 2]
```diff
@@ -33,108 +38,116 @@ for (package in azbatchenv$packages) {
...
+  parallel::clusterExport(cluster, "libPaths")
+  parallel::clusterEvalQ(cluster, .libPaths(libPaths))
+
+  doParallel::registerDoParallel(cluster)
 
   status <- tryCatch({
-    if (errorHandling == "remove" || errorHandling == "stop") {
+    count <- 1
+
+    # Verify if merge task is sub or final task
```
Comment (paselem) [inline inst/startup/merger.R]: Similar to before - should we be pulling this out to another change for multi-merge?

File: inst/startup/merger.R [hunk 3]
```diff
@@ -33,108 +38,116 @@ for (package in azbatchenv$packages) {
...
 
-      for (i in 1:batchTasksCount) {
-        taskPath <- file.path(
-          batchTaskWorkingDirectory,
-          "result",
-          paste0(i, "-result.rds")
-        )
+    results <- foreach::foreach(i = 1:length(files), .export = c("batchTaskWorkingDirectory",
+                                                                   "batchJobId",
+                                                                    "chunkSize",
```
Comment (paselem) [inline inst/startup/merger.R]: nitpick: spacing"

File: tests/testthat/test-error-handling.R [hunk 4]
```diff
@@ -0,0 +1,59 @@
+context("error handling test")
+test_that("Remove error handling test", {
```
Comment (paselem) [inline tests/testthat/test-error-handling.R]: Should we have one explicitly for 'stop'?

Comment (brnleehng) [reply tests/testthat/test-error-handling.R]: I'll add the test

File: R/batchHelperFunctions.R [hunk 5]
```diff
@@ -1,3 +1,182 @@
...
+    taskId,
+    environmentSettings = list(setting, containerEnv),
+    commandLine = commands,
+    dependsOn = dependsOn,
+    outputFiles = outputFiles,
+    exitConditions = exitConditions
+  )
+}
+
+addSubMergeTask <- function(jobId, taskId, rCommand, ...){
```
[the remark this answers is not in the document]
Comment (brnleehng) [reply R/batchHelperFunctions.R]: Pulling it out different PR

File: R/doAzureParallel.R [hunk 6]
```diff
@@ -223,6 +223,7 @@ setHttpTraffic <- function(value = FALSE) {
   assign("github", githubPackages, .doAzureBatchGlobals)
   assign("bioconductor", bioconductorPackages, .doAzureBatchGlobals)
   assign("pkgName", pkgName, .doAzureBatchGlobals)
+  assign("it", it, .doAzureBatchGlobals)
```
[the remark this answers is not in the document]
Comment (brnleehng) [reply R/doAzureParallel.R]: Removed

[the remark this answers is not in the document]
Comment (brnleehng) [reply R/doAzureParallel.R]: Previously testing it out on merge task to pass the iterator

File: R/doAzureParallel.R [hunk 7]
```diff
@@ -444,6 +451,23 @@ setHttpTraffic <- function(value = FALSE) {
     resourceFiles <-
       append(resourceFiles, requiredJobResourceFiles)
 
+    ntasks <- length(argsList)
+
+    startIndices <- seq(1, length(argsList), chunkSize)
+
+    endIndices <-
+      if (chunkSize >= length(argsList))
+      {
```
[the remark this answers is not in the document]
Comment (brnleehng) [reply R/doAzureParallel.R]: Fixed

File: inst/startup/merger.R (hunk 1 above)
Comment (brnleehng) [reply inst/startup/merger.R]: Fixed

File: R/doAzureParallel.R [hunk 8]
```diff
@@ -622,48 +686,53 @@ setHttpTraffic <- function(value = FALSE) {
...
+            # check for errors
+            errorValue <- foreach::getErrorValue(it)
+            errorIndex <- foreach::getErrorIndex(it)
+
+            if (identical(obj$errorHandling, "stop") &&
+                !is.null(errorValue)) {
+              msg <-
+                sprintf(
+                  paste0(
+                    "task %d failed - '%s'.\r\nBy default job and its result is deleted after run is over, use",
```
[the remark this answers is not in the document]
Comment (brnleehng) [reply R/doAzureParallel.R]: Fixed

File: inst/startup/merger.R (hunk 2 above)
Comment (brnleehng) [reply inst/startup/merger.R]: Yes removed

**Pull request merged**